### PR TITLE
Remove -Werror from autotools/darwin build, enable it on CI for all platforms

### PR DESCRIPTION
**Pull request by veprbl**
The `-Werror` was added very long time ago: c1e1d067d2f6b0fe184eae7ebf1fcb7b2d5ff5c7
This flag is more suitable for running on CI or during development, but encumbers downstream packaging.

cc NixOS/nixpkgs#136060

### CODE REVIEW

**Comment by Be-ing** [comment]
I don't think there's any good reason to remove this upstream. If packagers are annoyed by it, I think that's their problem and they can remove it if they want.

**Comment by veprbl** [comment]
@Be-ing Would you accept a patch that adds -Werror for autotools on linux and for cmake builds?

**Comment by Be-ing** [comment]
Thank you for noticing that inconsistency in the build system, which is one more argument for removing autotools. I think we should use `-Werror` with CMAKE_BUILD_TYPE=Debug.

**Comment by philburk** [comment]
We definitely want -Werror somewhere in our CI to catch warnings.

@veprbl - Do you have more details on the problems -Werror causes for packagers?

**Comment by r-burns** [comment]
Werror is good to enable for development, but unless your CI checks every possible compiler out there, warnings can always sneak by. For example, Nixpkgs ran into some new warnings with gcc 10. I think the recommended approach is to keep Werror off by default, and enable it on the CI using CPPFLAGS.

**Comment by veprbl** [comment]
> @veprbl - Do you have more details on the problems -Werror causes for packagers?

The compilation errors caused by `-Werror` are not so helpful to the end users. They are often seen as low importance and not reported upstream (One could say that upstream, in theory, has access to all the same compilers, so what is there to report). In the end work is put into silencing those errors without any productive output.

> We definitely want -Werror somewhere in our CI to catch warnings.

That is understandable. How about adding it in the CI workflows?

**Comment by philburk** [comment]
I am in favor of merging this after we add -Werror to the CI.

**Comment by philburk** [comment]
Note that the Werror broke the MSVC CI.

**Comment by veprbl** [comment]
> Note that the Werror broke the MSVC CI.

Correct. After using /WX it is still broken:
```
D:\a\portaudio\portaudio\build\asiosdk_2.3.3_2019-06-14\host\pc\asiolist.cpp(24,41): error C2220: the following warning is treated as an error [D:\a\portaudio\portaudio\build\PortAudio.vcxproj]
43
D:\a\portaudio\portaudio\build\asiosdk_2.3.3_2019-06-14\host\pc\asiolist.cpp(24,41): warning C4267: 'argument': conversion from 'size_t' to 'DWORD', possible loss of data [D:\a\portaudio\portaudio\build\PortAudio.vcxproj]
44
D:\a\portaudio\portaudio\build\asiosdk_2.3.3_2019-06-14\host\pc\asiolist.cpp(31,42): warning C4267: 'argument': conversion from 'size_t' to 'DWORD', possible loss of data [D:\a\portaudio\portaudio\build\PortAudio.vcxproj]
```

Any thoughts?

File: .github/workflows/autotools.yml [hunk 1]
```diff
@@ -11,6 +11,8 @@ jobs:
 
     runs-on: ubuntu-latest
     name: Ubuntu
+    env:
+      CXXFLAGS: "-Werror"
```
Comment (RossBencina) [inline .github/workflows/autotools.yml:15]: Should this not also set `CFLAGS` or, alternatively, perhaps only `CPPFLAGS`. I am not 100% sure how to address both C and C++, but most of our files are `C`. Some discussion here:

https://stackoverflow.com/questions/5541946/cflags-ccflags-cxxflags-what-exactly-do-these-variables-control

**Comment by RossBencina** [comment]
I think this is heading in the right direction. It's OK to limit `-Werror` to CI. Given that the problem with `/WX` on MSVC is with the ASIO SDK (something we can't fix), I think that we can move ahead with merging this without `/WX`.

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -19,6 +19,7 @@ jobs:
             cmake_generator: "Unix Makefiles"
             cmake_options:
               -DOSS=ON
+            werror_flags: "-Werror"
```
Comment (Be-ing) [inline .github/workflows/cmake.yml:22]: This should be done in CMakeLists.txt for `${CMAKE_BUILD_TYPE} STREQUAL "Debug"` and change the `cmake_build_type` variable in this GH Actions workflow file to `Debug`.

**Comment by Be-ing** [comment]
Please update the title of the pull request to reflect what it is actually doing now.

File: .github/workflows/autotools.yml (hunk 1 above)
Comment (veprbl) [reply .github/workflows/autotools.yml:16]: @RossBencina Right, if there is a C code, we should enable the flag there as well. `CPPFLAGS` would be the wrong flags, as CPP stands for C Pre Processor. I've implemented your suggestion using `CFLAGS` and `CXXFLAGS`.

File: .github/workflows/cmake.yml [hunk 3]
```diff
@@ -16,6 +16,8 @@ jobs:
             cmake_options:
               -DPA_USE_OSS=ON
               -DCMAKE_TOOLCHAIN_FILE=vcpkg/scripts/buildsystems/vcpkg.cmake
+              -DOSS=ON
```
Comment (RossBencina) [inline .github/workflows/cmake.yml:19]: What is the purpose of this `-DOSS=ON`? Is this needed. I'm not sure whether this is an unrelated change.

**Comment by RossBencina** [comment]
Regarding the warnings that are breaking CI, we already have tickets for these:
`jack_port_get_latency` deprecated: #641
`snd_pcm_sw_params_set_xfer_align` deprecated: #213

**Comment by philburk** [comment]
Thanks for adding the -Werror to the CI. They caused CI to fail on the deprecated functions, which is good.

We should fix the deprecated functions before we merge this PR.

File: .github/workflows/cmake.yml (hunk 3 above)
Comment (veprbl) [reply .github/workflows/cmake.yml]: Was an incorrect rebase on my side

**Comment by dechamps** [comment]
FYI I am proposing a slightly different approach in #780.
